### PR TITLE
fix(search): exempt paired facets calls from the SEC-S11 bucket

### LINKED ISSUE #1903: search: the facets request should not spend a rate-limit token when the results request already embedded the query
Since #1897, `/search/facets/` embeds the query so its counts cover the same candidate set as `/search/datasets/`, and both routes draw on one shared SEC-S11 rate-limit bucket (`semantic_search_rate_limit`, default 30 per minute per IP). The SPA fires the two requests as a pair on every query change, so one change spends two tokens while the provider is paid for one embedding (the second request is an embedding-cache hit when it arrives after the first, and a second provider call only when the pair overlaps). At the default an SPA user therefore gets 15 query changes a minute per IP, not 30.

The shared bucket is the constraint: #1897 rejected two buckets because a caller alternating the routes could embed twice the advertised cap, and rejected raising the default because a non-SPA client calling one route would then get double the cap the setting advertises. The fix has to keep one bucket and one cap and stop charging the facets half of the pair.

Options:

- Exempt the facets request from the bucket when the query embedding is already cached (`generate_embedding`'s TTL cache in `service_semantic.py`): a cache hit costs the provider nothing, so it should cost no token. slowapi's `exempt_when` is evaluated before the handler runs, so this needs the cache check hoisted into the limiter callable, and it does not help when the pair overlaps.
- Have the SPA request facets from the results response: one request per query change, facets computed in the same `select_candidates` call. This changes the `/search/datasets/` response shape (or adds an `include=facets` parameter) and the `useFacets` hook.

Either way the shared-bucket tests in `backend/tests/test_rate_limits.py` (`test_search_datasets_and_facets_share_one_bucket`) stay green: alternating novel queries must still hit one cap. Operators fronting many users behind one NAT can raise the stored `semantic_search_rate_limit` in the meantime.

### PR DESCRIPTION
**Pull request by ishiland**
## Summary

The search facets request spends a rate-limit token even when the results request already embedded the same query. Both `/search/datasets/` and `/search/facets/` draw on one shared SEC-S11 bucket (`semantic_search_rate_limit`, default 30/minute per IP), and the SPA fires them as an unordered pair on every query change (neither request awaits the other), so one query change spent two tokens while the provider was paid for at most one embedding. At the default, an SPA user got 15 query changes a minute per IP instead of 30.

Closes #1903.

## Changes

Both routes' `shared_limit` decorator now carries an `exempt_when` callable backed by a short-lived claim registry (`service_semantic.py`'s `_query_claims`, keyed on client + tenant + normalized query text):

- Whichever route's rate-limit gate is reached first for a given query claims it and pays the SEC-S11 token. The sibling route, arriving within a 5-second window, consumes that claim once and is exempt.
- Claiming is admission-gated: `exempt_when` only reads and consumes an existing claim (it runs before the limiter's own admit/reject check, so a request the bucket goes on to reject must not be able to seed one). A separate call at the top of each handler records a claim, which only runs for a request the rate limit actually let through.
- The claim key includes the same client identity the limiter's own `key_func` resolves, so one client can't consume another client's claim.
- A same-route repeat never matches its own claim, so a burst against one route still pays per request; a claim is single-use (deleted on consumption), not a standing exemption for the whole window.
- The registry is TTL-bound and hard-capped (256 entries, LRU eviction), so it can't grow unbounded even when a paired request's sibling never arrives.

`router.py`'s and `service_semantic.py`'s LOC caps in `test_layering.py` were re-measured and ratcheted at each step.

## Why this shape over the issue's other option

The issue also floated folding facets into the `/search/datasets/` response so the SPA makes one request instead of two. That removes the double spend outright, but it changes a published response shape for a saving the claim-registry approach gets without touching the wire contract.

## Review history

Four codex rounds, each catching a real gap in the exemption mechanism:
- **Round 1**: the first design assumed the results call always embeds first and checked the embedding cache; the two requests are actually unordered, so the exemption rarely fired. Replaced with the claim registry, symmetric on both routes.
- **Round 2**: the claim wasn't consumed, so a same-route burst against one query rode free for the whole TTL window even though each request still billed the provider. Made claiming single-use and cross-route only.
- **Round 3**: two findings. A rejected request could still seed a claim a follow-up call redeemed, fixed by admission-gating. The claim wasn't scoped to the caller, so a different client could consume it, fixed by including the limiter's client identity in the key.
- **Round 4**: the global per-IP default was being charged twice per request. Both routes' rate limit is a callable value, which slowapi registers differently from a static one; its ASGI middleware doesn't recognize that registration and runs its own check first, always charging the global default for these routes regardless of `override_defaults`. I'd added that setting believing an exempted call would otherwise have no limit at all, which isn't true for this kind of route, so I removed it. Round 4 also flagged that the claim registry is process-local, so a paired request split across the two Uvicorn workers the bundled prod compose file runs by default gets no exemption. That's an existing, explicitly-deferred characteristic of the rate limiter itself (see `query_router.py`'s `fix(#565)` note): the worst case is the pre-#1903 behavior, never worse, so it's documented rather than fixed here. Filed #2018 to track a cross-worker-aware limiter/registry.

Fixing round 4's global-limit bug surfaced a gate-coverage gap: `test_semantic_search_rate_limit_1778.py` structurally asserts the route's `override_defaults` stays `True`, and the round-1 change had been silently failing it since the first round, because that test file doesn't mention "facets" and wasn't caught by this lane's facets-grep-based test search. It's green again, and the final test sweep covers every test file in `backend/tests/` that references the touched routes, functions, or symbols, not just the facets-adjacent ones.

- **Round 5**: two more findings. A rate-limit exemption can be granted before either half of a paired request finishes embedding, so a genuinely overlapping pair (both missing the embedding TTL cache, which is the common case for any novel query since coordination happens at the rate-limit gate) could still make two paid provider calls under one charged token. Concurrent calls for the same query now join a single in-flight embedding task instead of each starting their own. Separately, `tenant_cache_key()` raises when multi-tenant mode has no verified tenant context; the claim key builder called it unconditionally, so a search query on a trusted unscoped host would 500. Both fixed.
- **Round 6** (independent adversarial review, verdict: merge, no P1): live instrumentation of the gate confirmed the counting invariant (`exempt_when` fires exactly once per request, an exempted request records nothing, a 429'd request records nothing), and traced two P2s and several P3s in the round-5 single-flight, the only code no automated round had seen yet.
  - The shared embedding task had no cancellation isolation: awaiting a bare `Task` makes it the awaiter's `_fut_waiter`, so cancelling any one participant cancelled the shared task and every other participant with it, and `CancelledError` isn't caught by `resolve_semantic_arm`'s `except Exception`. Every participant now awaits `asyncio.shield(task)`. That raised a session-safety question: a shielded task can outlive the request that spawned it, and `generate_embeddings_batch` reads the session for a default when `dimensions` is unpinned. The single-flight now only applies when the config is fully pinned; an unpinned call runs standalone, as before round 5.
  - The test fixture cleared the embedding cache but not the new in-flight registry; a leaked task bound to a closed event loop could poison a later test. Fixed.
  - The claim key joined the client address and query text with `:`, which an IPv6 client address also contains, so a /64 holder could pick low bits that collide with a neighbour's key. The key is now a tuple, not a string join.
  - The route tag was a substring test against the URL path (`"facets" in request.url.path`), correct today but fragile against a future route sharing that text. It now reads the route's own name from `request.scope["route"]`.
  - After the reviewer verified those four fixes, one more surfaced: the single-flight's fully-pinned gate tested only `dimensions is None`, but the session-reading condition it needs to match is `not model or not dimensions` in `generate_embeddings_batch`; an empty `embedding_model` string passes config validation with no non-empty check, so that case would still have let a shared task read the originator's session. The gate now tests the exact predicate it's guarding against.

## Known limitations

- The claim registry (and the SEC-S11 rate limiter it sits on) is per-process. Under a multi-worker deployment, a paired request split across workers is not exempted: each worker charges its own token, same as before this fix. Tracked at #2018.
- A query under `_MIN_SEMANTIC_QUERY_LEN` (4 characters) never reaches the embedding provider, but the claim/exemption mechanism still applies to it, so the effective request cap for that non-semantic path doubles even though provider spend on it stays at zero.
- A request that joins an in-flight embed inherits the originator's remaining `_QUERY_EMBED_TIMEOUT_SECONDS` (8s) deadline, not a fresh one. A late joiner can time out and fall back to full-text search sooner than an unpaired request would.

## Area

- [x] Backend (API, services, models)

## Testing

- [x] Unit tests pass (`pytest`)
  - `test_rate_limits.py`: 6 new tests, each with a counterfactual verified against the commit it was written to catch. The paired-request fix (both orderings), the same-route-burst fix, the rejected-request fix, the cross-client fix (spins up a second `ASGITransport` client at a different IP against the same app instance), the tenant-unscoped fix, and a four-request `datasets, facets, datasets, facets` chain pinning that it spends exactly two tokens.
  - `test_search_embedding_cache.py`: 3 new tests for the round-5/6 single-flight, including one asserting `len(_embedding_inflight) == 1` while two calls are joined (pins the mechanism, not a side effect of a mock that never yields) and one proving a cancelled participant doesn't take a sibling down with it.
  - `test_layering.py`, `test_rule1_structural.py`, `test_semantic_search_rate_limit_1778.py` (the pre-existing test the round-4 fix restores)
  - `test_search_facets.py`, `test_search_facets_input_cap.py`, `test_search_cache.py`, `test_hybrid_search.py`, `test_datasets.py`, `test_search_datetime.py`, `test_redirect_slashes.py`, `test_basemap_sublayer_overrides.py`, `test_cache.py`, `test_ogc_features_filter.py`, `test_sdks_round_trip.py`, `test_search_simple_regconfig.py`, `test_search_cors_1596.py`
  - `ruff check` / `ruff format --check`, `finding_markers.py`
- [ ] Manual testing (describe below)
- [ ] Playwright e2e (if UI changes)
- [ ] No tests needed (docs, config, etc.)

No route signature, query parameter, or published description changed, so no OpenAPI/SDK regen is needed.

## Checklist

- [x] Code follows existing project conventions
- [ ] i18n: new user-facing strings added to all 4 locales (en, fr, es, de)
- [x] No new lint warnings (`ruff check`, `eslint`)
- [x] TypeScript compiles without errors
- [ ] Migration included (if DB schema changed)
- [x] No secrets or credentials in the diff
- [ ] Release/docs deploys: ran `make deployed-surface-check` after the marketing/docs deploy and recorded the result in Testing

### CODE REVIEW

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/search/router.py [hunk 1]
```diff
@@ -422,7 +433,15 @@ def _semantic_search_rate_limit(_request: Request | None = None) -> str:
...
 # fix(#1855): facets embed the query, so both search routes draw on ONE SEC-S11
 # bucket; two buckets let a caller alternating them embed twice the cap.
-@limiter.shared_limit(_semantic_search_rate_limit, scope="semantic_search")
+# fix(#1903): override_defaults=False keeps the global per-IP default active
+# on an exempted call -- without it, an exempt Limit leaves NO limit at all
+# for that request (slowapi drops the defaults whenever a route limit wins).
+@limiter.shared_limit(
+    _semantic_search_rate_limit,
+    scope="semantic_search",
+    exempt_when=_facets_query_embedding_cached,
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/router.py]: Fixed by 689a32d80: the exemption no longer depends on the results call embedding first. Both routes now run the same `exempt_when` (`claim_semantic_search_query` in `service_semantic.py`), which claims the query for whichever route's rate-limit gate reaches it first and exempts the other — no dependency on the SPA's request ordering, and no dependency on the embed actually finishing. Verified with a targeted counterfactual (stripping `exempt_when` from just the results route) that both pairing orders regress without the symmetric fix; both orders now pass in `test_paired_query_claims_the_bucket_once_whichever_route_is_first`.

File: backend/app/modules/catalog/search/service_semantic.py [hunk 2]
```diff
@@ -78,6 +78,26 @@ def _embedding_cache_clear() -> None:
...
+    lookup on the full tuple, so it can never serve a wrong vector.
+    """
+    normalized = text.strip().lower()
+    if not normalized:
+        return False
+    prefix = tenant_cache_key(normalized)
+    now = time.monotonic()
+    return any(
+        key[0] == prefix and expires_at >= now
+        for key, (expires_at, _vector) in _embedding_cache.items()
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Fixed by 689a32d80: the exemption no longer reads the embedding cache at all, so a stale model/fingerprint match can't happen. The new `claim_semantic_search_query` registry keys only on tenant + normalized query text with its own short (5s) TTL, decoupled from the embedding cache's config-scoped keys and its 300s TTL. A config change can no longer create a window where a request is exempted from the rate limit but still bills the provider.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/search/service_semantic.py [hunk 3]
```diff
@@ -78,6 +78,48 @@ def _embedding_cache_clear() -> None:
...
+    reach its gate claims the query, the second (whichever route) is exempt.
+    """
+    normalized = text.strip().lower()
+    if not normalized:
+        return False
+    key = tenant_cache_key(normalized)
+    now = time.monotonic()
+    expires_at = _query_claims.get(key)
+    if expires_at is not None and expires_at >= now:
+        return True
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Fixed by 5109ca1f5: the claim now stores which route made it and is deleted the moment the OTHER route consumes it (`claim_semantic_search_query(text, route)` in `service_semantic.py`). A same-route repeat never matches its own claim, so a burst against one route pays per request as before; only the genuine cross-route pair gets the single exemption. Added `test_same_route_repeat_does_not_ride_a_cross_route_claim`, verified it fails against the prior commit (third same-route call returned 200 instead of the expected 429) and passes against this one.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/search/service_semantic.py [hunk 4]
```diff
@@ -78,6 +78,60 @@ def _embedding_cache_clear() -> None:
...
+    now = time.monotonic()
+    claimed = _query_claims.get(key)
+    if claimed is not None:
+        claimant_route, expires_at = claimed
+        if expires_at >= now:
+            if claimant_route == route:
+                return False
+            del _query_claims[key]
+            return True
+    _query_claims[key] = (route, now + _QUERY_CLAIM_TTL_SECONDS)
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Fixed by 2aa010e32: claiming is now split into a read-only consume (`consume_paired_query_claim`, called from `exempt_when` before the limiter checks the bucket) and a record step (`record_paired_query_claim`, called from `_finalize_semantic_search_claim` at the top of each handler). A request the bucket rejects never reaches the handler body, so it can no longer seed a claim. Added `test_rejected_request_does_not_seed_a_claim`, verified it fails against the prior commit (the facets call after a rejected datasets call for the same query returned 200) and passes against this one.

File: backend/app/modules/catalog/search/service_semantic.py [hunk 5]
```diff
@@ -78,6 +78,60 @@ def _embedding_cache_clear() -> None:
...
+    Called from both search routes' rate-limit ``exempt_when`` hooks, which
+    run before either handler body -- so the first of a paired request to
+    reach its gate claims the query for its own route, and only the sibling
+    route's request, arriving within the window, is exempt (the claim is
+    then deleted, so this is a one-time consume, not a standing exemption).
+    """
+    normalized = text.strip().lower()
+    if not normalized:
+        return False
+    key = tenant_cache_key(normalized)
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Fixed by 2aa010e32: the claim key now includes `get_remote_address(request)`, the same identity the limiter's own `key_func` uses to scope the SEC-S11 bucket. A different client can no longer consume this client's claim. Added `test_claim_is_scoped_to_the_requesting_client` (two separate ASGITransport clients against the same app instance), verified it fails against the prior commit (client B's follow-up call showed an untouched bucket) and passes against this one.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/search/router.py [hunk 6]
```diff
@@ -422,7 +463,15 @@ def _semantic_search_rate_limit(_request: Request | None = None) -> str:
...
 # bucket; two buckets let a caller alternating them embed twice the cap.
-@limiter.shared_limit(_semantic_search_rate_limit, scope="semantic_search")
+# fix(#1903): override_defaults=False keeps the global per-IP default active
+# on an exempted call -- without it, an exempt Limit leaves NO limit at all
+# for that request (slowapi drops the defaults whenever a route limit wins).
+@limiter.shared_limit(
+    _semantic_search_rate_limit,
+    scope="semantic_search",
+    exempt_when=_semantic_search_query_already_claimed,
+    override_defaults=False,
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/router.py]: Fixed by 00b7fcfe7: dropped override_defaults=False from both routes. Traced it in slowapi's source: this route's limit_value is callable (admin-editable at runtime), so slowapi files it under _dynamic_route_limits rather than _route_limits, and SlowAPIMiddleware._should_exempt() only checks the latter. The middleware therefore runs its own _check_request_limit before the endpoint at all, which for these routes unconditionally adds the global default (dynamic limits are invisible to that call, so the empty route_limits list makes combined_defaults vacuously True) -- regardless of override_defaults. That's an existing, pre-#1903 characteristic (test_semantic_search_rate_limit_1778.py's test_semantic_search_rate_limit_decorator_present already asserts override_defaults stays True structurally for this reason), which override_defaults=False was undermining by adding a second, redundant charge in the decorator's own check. That test file wasn't in this lane's search-facets-related grep and had actually been failing since the first round; it's green again now, along with a broader sweep of every test referencing the affected symbols.

File: backend/app/modules/catalog/search/service_semantic.py [hunk 7]
```diff
@@ -78,6 +78,71 @@ def _embedding_cache_clear() -> None:
...
+# request for the same popular query consume this client's claim and skip
+# its own bucket. Exempting is a single-use, cross-route consume, not a
+# standing amnesty: a same-route repeat never matches its own claim, and
+# the opposite route's exemption deletes it, so a third request pays fresh.
+# Deliberately much shorter than the embedding cache TTL above, and it
+# never looks at that cache, so a config change mid-window cannot make a
+# stale-model match exempt a call that ends up billing the provider.
+_QUERY_CLAIM_TTL_SECONDS = 5.0
+_QUERY_CLAIM_MAX_SIZE = 256
+_query_claims: "OrderedDict[str, tuple[str, float]]" = OrderedDict()
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Confirmed and documented in 13014d5da rather than fixed: the claim registry is per-process, same as the SEC-S11 bucket it exempts from (app.platform.ratelimit's Limiter uses slowapi's default in-memory storage, no storage_uri configured). Under the bundled multi-worker prod config, a paired request split across two workers gets no exemption -- each worker pays its own token, which is exactly today's pre-#1903 behavior, never worse. Sharing this (or the rate limiter itself, which already has the same characteristic per query_router.py's fix(#565) note: "a Valkey-backed limiter is a tracked app-wide change, not forked here") needs a distributed store. Given that's an existing, explicitly-deferred, cross-cutting decision in this codebase, I flagged it to my lead rather than expanding this fix's scope to include it unilaterally.

[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Follow-up after review with my lead: leaving this documented, not fixed, in this PR. The degraded case (a paired request split across workers pays two tokens) is exactly the pre-#1903 behavior, never worse, and the underlying rate limiter already has the identical per-process characteristic (query_router.py's fix(#565) note calls a Valkey-backed limiter "a tracked app-wide change, not forked here" for the same reason). Confirmed and documented in 539563fb2 that this can't become new attack surface: the registry is TTL-bound (5s) and hard-capped at 256 entries via LRU eviction, and consuming a claim deletes it (single-use), so an unconsumed claim from a sibling that landed on another worker can't grow the dict or be redeemed more than once. Filed #2018 to track making the rate limiter and this registry cross-worker aware, and cited it from the registry's comment.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/search/router.py [hunk 8]
```diff
@@ -422,7 +463,19 @@ def _semantic_search_rate_limit(_request: Request | None = None) -> str:
...
+# _dynamic_route_limits rather than _route_limits. SlowAPIMiddleware's
+# _should_exempt() only checks _route_limits, so it does not recognize this
+# route as decorator-handled and runs its OWN check first, always charging
+# the global default regardless of override_defaults -- an exempted call is
+# still bounded by that middleware-level charge. override_defaults=False
+# would only add a SECOND, redundant global-default charge here.
+@limiter.shared_limit(
+    _semantic_search_rate_limit,
+    scope="semantic_search",
+    exempt_when=_semantic_search_query_already_claimed,
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/router.py:468]: Fixed by bee1bdc73: concurrent calls to generate_embedding() for the same (text, model, fingerprint) now join a single in-flight asyncio.Task instead of each starting their own provider call. Only the originating call's session does any I/O; a caller that joins an existing task never touches it, so no AsyncSession is shared across coroutines. Added test_concurrent_calls_for_the_same_key_coalesce_into_one_provider_call, which uses asyncio.Event (not sleeps) to provably start the second call while the first is still blocked inside the mocked provider call, matching the genuine race rather than a lucky timing. Verified it fails against the prior commit (call_count == 2) and passes against this one.

File: backend/app/modules/catalog/search/service_semantic.py [hunk 9]
```diff
@@ -78,6 +78,85 @@ def _embedding_cache_clear() -> None:
...
+
+
+def _query_claim_key(client_key: str, text: str) -> str | None:
+    normalized = text.strip().lower()
+    if not normalized:
+        return None
+    return f"{client_key}:{tenant_cache_key(normalized)}"
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/search/service_semantic.py]: Fixed by bee1bdc73: _query_claim_key now checks tenant_cache_context_available() before calling tenant_cache_key(), the same guard the search cache already uses, so an unscoped multi-tenant request disables claiming instead of raising. Added test_claim_functions_degrade_without_crashing_when_tenant_unscoped, verified it fails against the prior commit (ValueError) and passes against this one.
